File: frontend/src/services/video/analyses.ts
// @ts-ignore
/* eslint-disable */
import { request, type RequestOptions } from "@/lib/request";

/** 查询视频分析任务 查询分析进度及经过证据校验的结果。 GET /api/analyses/${param0} */
export async function getAnalysis(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getAnalysisParams,
  options?: RequestOptions
) {
  const { analysis_id: param0, ...queryParams } = params;
  return request<API.AnalysisResponse>(`/api/analyses/${param0}`, {
    method: "GET",
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** 删除视频分析与报告 隐藏分析任务并异步清理其私有报告对象。 DELETE /api/analyses/${param0} */
export async function deleteAnalysis(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.deleteAnalysisParams,
  options?: RequestOptions
) {
  const { analysis_id: param0, ...queryParams } = params;
  return request<any>(`/api/analyses/${param0}`, {
    method: "DELETE",
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** 取消视频分析任务 请求取消尚未结束的视频分析任务。 POST /api/analyses/${param0}/cancel */
export async function cancelAnalysis(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.cancelAnalysisParams,
  options?: RequestOptions
) {
  const { analysis_id: param0, ...queryParams } = params;
  return request<API.AnalysisResponse>(`/api/analyses/${param0}/cancel`, {
    method: "POST",
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** 导出视频分析报告 将已完成的结构化分析结果导出为 DOCX 报告。 GET /api/analyses/${param0}/report.docx */
export async function exportAnalysisReport(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.exportAnalysisReportParams,
  options?: RequestOptions
) {
  const { analysis_id: param0, ...queryParams } = params;
  return request<string>(`/api/analyses/${param0}/report.docx`, {
    method: "GET",
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** 导出 Markdown 视频分析报告 导出与前端预览、DOCX 转换共用的唯一 Markdown 报告。 GET /api/analyses/${param0}/report.md */
export async function exportAnalysisMarkdown(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.exportAnalysisMarkdownParams,
  options?: RequestOptions
) {
  const { analysis_id: param0, ...queryParams } = params;
  return request<string>(`/api/analyses/${param0}/report.md`, {
    method: "GET",
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** 重试原视频分析任务 为同一分析任务创建下一执行代次，不改变任务资源 ID。

Retry 是上一运行的无参数重放；带请求体的请求按校验错误拒绝。 POST /api/analyses/${param0}/retry */
export async function retryAnalysis(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.retryAnalysisParams,
  options?: RequestOptions
) {
  const { analysis_id: param0, ...queryParams } = params;
  return request<API.AnalysisResponse>(`/api/analyses/${param0}/retry`, {
    method: "POST",
    params: { ...queryParams },
    ...(options || {}),
  });
}

/** 列出输入兼容的分析 Skill 按输入类型返回可选 Skill 及用户可编辑的默认提示词。 GET /api/analysis-skills */
export async function listAnalysisSkills(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listAnalysisSkillsParams,
  options?: RequestOptions
) {
  return request<API.AnalysisSkillResponse[]>("/api/analysis-skills", {
    method: "GET",
    params: {
      ...params,
    },
    ...(options || {}),
  });
}

/** 创建视频分析任务 基于已完成的下载制品创建异步 AI 分析任务。 POST /api/downloads/${param0}/analyses */
export async function createAnalysis(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.createAnalysisParams,
  body: API.AnalysisRequest,
  options?: RequestOptions
) {
  const { download_id: param0, ...queryParams } = params;
  return request<API.AnalysisResponse>(`/api/downloads/${param0}/analyses`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    params: { ...queryParams },
    data: body,
    ...(options || {}),
  });
}

/** 读取下载任务最近的视频分析 恢复当前用户在该下载任务上最近创建的分析与报告。 GET /api/downloads/${param0}/analysis */
export async function getLatestDownloadAnalysis(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getLatestDownloadAnalysisParams,
  options?: RequestOptions
) {
  const { download_id: param0, ...queryParams } = params;
  return request<API.AnalysisResponse | null>(
    `/api/downloads/${param0}/analysis`,
    {
      method: "GET",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

File: backend/app/api/routes/analyses.py
from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Request, Response, status
from fastapi.exceptions import RequestValidationError

from app.api.admission import RateLimitAdmission
from app.api.deps import IdempotencyKey, get_analysis_use_cases, get_current_user
from app.api.responses import ApiResponseRoute
from app.core.runtime import AnalysisUseCases
from app.schemas.analyses import (
    AnalysisRequest,
    AnalysisResponse,
    AnalysisSkillResponse,
)
from app.services.analysis.export_report import DOCX_MEDIA_TYPE, MARKDOWN_MEDIA_TYPE
from app.services.analysis.rules.enums import AnalysisInputKind
from app.services.auth.models import CurrentUser

router = APIRouter(route_class=ApiResponseRoute, tags=["analyses"])
User = Annotated[CurrentUser, Depends(get_current_user)]
UseCases = Annotated[AnalysisUseCases, Depends(get_analysis_use_cases)]


@router.get(
    "/analysis-skills",
    operation_id="listAnalysisSkills",
    response_model=tuple[AnalysisSkillResponse, ...],
    summary="列出输入兼容的分析 Skill",
)
async def list_analysis_skills(
    use_cases: UseCases,
    input_kind: AnalysisInputKind,
) -> tuple[AnalysisSkillResponse, ...]:
    """按输入类型返回可选 Skill 及用户可编辑的默认提示词。"""
    return tuple(
        AnalysisSkillResponse.model_validate(skill)
        for skill in use_cases.list_analysis_skills(input_kind)
    )


@router.post(
    "/downloads/{download_id}/analyses",
    operation_id="createAnalysis",
    dependencies=[Depends(RateLimitAdmission("analysis"))],
    response_model=AnalysisResponse,
    status_code=status.HTTP_201_CREATED,
    summary="创建视频分析任务",
)
async def create_analysis(
    download_id: UUID,
    body: AnalysisRequest,
    idempotency_key: IdempotencyKey,
    response: Response,
    user: User,
    use_cases: UseCases,
) -> AnalysisResponse:
    """基于已完成的下载制品创建异步 AI 分析任务。"""
    view = await use_cases.create_analysis(
        download_id,
        user.owner_hash,
        idempotency_key,
        body.skill_id,
        body.output_language,
        body.custom_prompt,
        quota=user.admission_quota,
    )
    response.headers["Location"] = f"/api/analyses/{view.id}"
    return AnalysisResponse.from_view(view)


@router.get(
    "/downloads/{download_id}/analysis",
    operation_id="getLatestDownloadAnalysis",
    response_model=AnalysisResponse | None,
    summary="读取下载任务最近的视频分析",
)
async def get_latest_download_analysis(
    download_id: UUID,
    user: User,
    use_cases: UseCases,
) -> AnalysisResponse | None:
    """恢复当前用户在该下载任务上最近创建的分析与报告。"""
    view = await use_cases.get_latest_download_analysis(download_id, user.owner_hash)
    return None if view is None else AnalysisResponse.from_view(view)


@router.get(
    "/analyses/{analysis_id}",
    operation_id="getAnalysis",
    response_model=AnalysisResponse,
    summary="查询视频分析任务",
)
async def get_analysis(
    analysis_id: UUID,
    user: User,
    use_cases: UseCases,
) -> AnalysisResponse:
    """查询分析进度及经过证据校验的结果。"""
    view = await use_cases.get_analysis(analysis_id, user.owner_hash)
    return AnalysisResponse.from_view(view)


@router.get(
    "/analyses/{analysis_id}/report.md",
    operation_id="exportAnalysisMarkdown",
    response_class=Response,
    responses={
        200: {
            "description": "Canonical Markdown analysis report",
            "content": {
                "text/markdown": {"schema": {"type": "string", "format": "binary"}}
            },
        }
    },
    summary="导出 Markdown 视频分析报告",
)
async def export_analysis_markdown(
    analysis_id: UUID,
    user: User,
    use_cases: UseCases,
) -> Response:
    """导出与前端预览、DOCX 转换共用的唯一 Markdown 报告。"""
    report = await use_cases.export_analysis_markdown(analysis_id, user.owner_hash)
    return Response(
        content=report.content,
        media_type=MARKDOWN_MEDIA_TYPE,
        headers={
            "Content-Disposition": f'attachment; filename="{report.filename}"',
            "X-Content-Type-Options": "nosniff",
        },
    )


@router.get(
    "/analyses/{analysis_id}/report.docx",
    operation_id="exportAnalysisReport",
    response_class=Response,
    responses={
        200: {
            "description": "Microsoft Word analysis report",
            "content": {
                DOCX_MEDIA_TYPE: {"schema": {"type": "string", "format": "binary"}}
            },
        }
    },
    summary="导出视频分析报告",
)
async def export_analysis_report(
    analysis_id: UUID,
    user: User,
    use_cases: UseCases,
) -> Response:
    """将已完成的结构化分析结果导出为 DOCX 报告。"""
    report = await use_cases.export_analysis_report(analysis_id, user.owner_hash)
    return Response(
        content=report.content,
        media_type=report.media_type,
        headers={
            "Content-Disposition": f'attachment; filename="{report.filename}"',
            "X-Content-Type-Options": "nosniff",
        },
    )


@router.post(
    "/analyses/{analysis_id}/cancel",
    operation_id="cancelAnalysis",
    response_model=AnalysisResponse,
    summary="取消视频分析任务",
)
async def cancel_analysis(
    analysis_id: UUID,
    user: User,
    use_cases: UseCases,
) -> AnalysisResponse:
    """请求取消尚未结束的视频分析任务。"""
    view = await use_cases.cancel_analysis(analysis_id, user.owner_hash)
    return AnalysisResponse.from_view(view)


@router.post(
    "/analyses/{analysis_id}/retry",
    operation_id="retryAnalysis",
    dependencies=[Depends(RateLimitAdmission("analysis_retry"))],
    response_model=AnalysisResponse,
    status_code=status.HTTP_201_CREATED,
    summary="重试原视频分析任务",
)
async def retry_analysis(
    analysis_id: UUID,
    request: Request,
    idempotency_key: IdempotencyKey,
    response: Response,
    user: User,
    use_cases: UseCases,
) -> AnalysisResponse:
    """为同一分析任务创建下一执行代次，不改变任务资源 ID。

    Retry 是上一运行的无参数重放；带请求体的请求按校验错误拒绝。
    """
    if (await request.body()).strip() not in {b"", b"null"}:
        raise RequestValidationError(
            errors=[
                {
                    "type": "extra_forbidden",
                    "loc": ("body",),
                    "msg": "Retry does not accept a request body.",
                    "input": None,
                }
            ]
        )
    view = await use_cases.retry_analysis(
        analysis_id,
        user.owner_hash,
        idempotency_key,
        quota=user.admission_quota,
    )
    response.headers["Location"] = f"/api/analyses/{view.id}"
    return AnalysisResponse.from_view(view)


@router.delete(
    "/analyses/{analysis_id}",
    operation_id="deleteAnalysis",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="删除视频分析与报告",
)
async def delete_analysis(
    analysis_id: UUID,
    user: User,
    use_cases: UseCases,
) -> Response:
    """隐藏分析任务并异步清理其私有报告对象。"""
    await use_cases.delete_analysis(analysis_id, user.owner_hash)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
